src/SharedStream.mo

```diff
@@ -0,0 +1,171 @@
+import Iter "mo:base/Iter";
+import Error "mo:base/Error";
+
+module {
+
+  type List<T> = ?{ e : T; var next : List<T> };
+
+  // reference implementation, not tested
+  class TemporaryQueueImpl<T>() {
+    var list : List<T> = null;
+
+    var histCnt_ : Nat = 0;
+    var headCnt_ : Nat = 0;
+    var tailCnt_ : Nat = 0;
+
+    var hist : List<T> = null;
+    var head : List<T> = null;
+    var tail : List<T> = null;
+
+    public func headId() : Nat = headCnt_;
+
+    public func size() : Nat = tailCnt_ - headCnt_;
+    public func historySize() : Nat = headCnt_ - histCnt_;
+    public func fullSize() : Nat = headCnt_ - histCnt_;
+
+    public func push(elements : [T]) {
+      for (el in elements.vals()) {
+        let oldTail = tail;
+        tail := ?{ e = el; var next = null };
+        switch (oldTail) {
+          case (?t) t.next := tail;
+          case (null) {
+            head := tail;
+            hist := tail;
+          };
+        };
+      };
+      tailCnt_ += elements.size();
+    };
+
+    public func pop() : ?T {
+      switch (head) {
+        case (?h) {
+          head := h.next;
+          switch (head) {
+            case (null) tail := null;
+            case (_) {};
+          };
+          headCnt_ += 1;
+          ?h.e;
+        };
+        case (null) { null };
+      };
+    };
+
+    public func peek() : ?T {
+      switch (head) {
+        case (?h) ?h.e;
+        case (null) null;
+      };
+    };
+
+    public func restore() {
+      head := hist;
+      headCnt_ := histCnt_;
+    };
```

Comment:
What we need is a function like this:
```
public func restore(headId : Nat) {
  if (headId < history) { Debug.trap("cannot happen") };
  head := min(head, headId);
}
```
This will be called from the `catch` branch of `emit`.